File: lib/alert-evaluator.ts
import { computeEma, computeRsi, computeSma, computeStochastic, enrichBars } from "./indicator-engine.js";
import type { AlertRow, EnrichedBar, EvaluationResult, OhlcvBar } from "./types.js";

function emaCross(
  prevFast: number,
  prevSlow: number,
  fast: number,
  slow: number,
  direction: "up" | "down",
): boolean {
  if (direction === "up") return prevFast <= prevSlow && fast > slow;
  return prevFast >= prevSlow && fast < slow;
}

function priceMaCross(
  prevClose: number,
  prevMa: number,
  close: number,
  ma: number,
  direction: "up" | "down",
): boolean {
  if (direction === "up") return prevClose <= prevMa && close > ma;
  return prevClose >= prevMa && close < ma;
}

function rsiThreshold(value: number, threshold: number, operator: "<" | ">"): boolean {
  return operator === "<" ? value < threshold : value > threshold;
}

function priceLevelCross(
  prevClose: number,
  close: number,
  level: number,
  operator: ">=" | "<=",
): boolean {
  if (operator === ">=") return prevClose < level && close >= level;
  return prevClose > level && close <= level;
}

function priceRangeBreakout(
  prevClose: number,
  close: number,
  low: number,
  high: number,
): boolean {
  const brokeUp = priceLevelCross(prevClose, close, high, ">=");
  const brokeDown = priceLevelCross(prevClose, close, low, "<=");
  return brokeUp || brokeDown;
}

function resolveRsiValue(
  current: EnrichedBar,
  enriched: EnrichedBar[],
  period: number,
): number {
  const rsiKey = `rsi_${period}` as keyof EnrichedBar;
  let rsiVal = current[rsiKey] as number | undefined;
  if (rsiVal === undefined) {
    rsiVal = computeRsi(enriched.map((b) => b.close), period).at(-1)!;
  }
  return rsiVal;
}

function resolveStochasticValue(enriched: EnrichedBar[], period: number): number {
  const stoch = computeStochastic(
    enriched.map((b) => b.high),
    enriched.map((b) => b.low),
    enriched.map((b) => b.close),
    period,
  );
  return stoch.at(-1)!;
}

function evaluateRsiPreset(
  preset: "rsi_oversold" | "rsi_overbought",
  params: Record<string, unknown>,
  current: EnrichedBar,
  enriched: EnrichedBar[],
): boolean {
  const period = Number(params.period ?? 14);
  const defaultThreshold = preset === "rsi_oversold" ? 30 : 70;
  const threshold = Number(params.threshold ?? defaultThreshold);
  const operator = preset === "rsi_oversold" ? "<" : ">";
  const rsiVal = resolveRsiValue(current, enriched, period);
  return rsiThreshold(rsiVal, threshold, operator);
}

function evaluateStochPreset(
  preset: "stoch_oversold" | "stoch_overbought",
  params: Record<string, unknown>,
  enriched: EnrichedBar[],
): boolean {
  const period = Number(params.period ?? 7);
  const defaultThreshold = preset === "stoch_oversold" ? 20 : 80;
  const threshold = Number(params.threshold ?? defaultThreshold);
  const operator = preset === "stoch_oversold" ? "<" : ">";
  const stochVal = resolveStochasticValue(enriched, period);
  if (!Number.isFinite(stochVal)) return false;
  return rsiThreshold(stochVal, threshold, operator);
}

function evaluateCustom(
  params: Record<string, unknown>,
  enriched: EnrichedBar[],
): boolean {
  const current = enriched.at(-1)!;
  const previous = enriched.at(-2)!;
  const type = params.type;

  if (type === "ema") {
    const fastLen = Number(params.ema_fast);
    const slowLen = Number(params.ema_slow);
    const direction = params.direction as "up" | "down";
    const closes = enriched.map((b) => b.close);
    const fastKey = `ema_${fastLen}` as keyof EnrichedBar;
    const slowKey = `ema_${slowLen}` as keyof EnrichedBar;
    let fastSeries = enriched.map((b) => b[fastKey] as number | undefined);
    let slowSeries = enriched.map((b) => b[slowKey] as number | undefined);
    if (fastSeries.some((v) => v === undefined)) {
      fastSeries = computeEma(closes, fastLen);
    }
    if (slowSeries.some((v) => v === undefined)) {
      slowSeries = computeEma(closes, slowLen);
    }
    return emaCross(
      fastSeries.at(-2)!,
      slowSeries.at(-2)!,
      fastSeries.at(-1)!,
      slowSeries.at(-1)!,
      direction,
    );
  }

  if (type === "price_ma") {
    const period = Number(params.period);
    const direction = params.direction as "up" | "down";
    const closes = enriched.map((b) => b.close);
    const maSeries =
      params.ma_type === "ema" ? computeEma(closes, period) : computeSma(closes, period);
    const i = closes.length - 1;
    const prevI = i - 1;
    const ma = maSeries[i];
    const prevMa = maSeries[prevI];
    if (!Number.isFinite(ma) || !Number.isFinite(prevMa)) {
      return false;
    }
    return priceMaCross(closes[prevI], prevMa, closes[i], ma, direction);
  }

  if (type === "rsi") {
    const period = Number(params.period ?? 14);
    const rsiVal = resolveRsiValue(current, enriched, period);
    return rsiThreshold(
      rsiVal,
      Number(params.threshold),
      params.operator as "<" | ">",
    );
  }

  if (type === "stochastic") {
    const period = Number(params.period ?? 7);
    const stochVal = resolveStochasticValue(enriched, period);
    if (!Number.isFinite(stochVal)) return false;
    return rsiThreshold(
      stochVal,
      Number(params.threshold),
      params.operator as "<" | ">",
    );
  }

  if (type === "price_level") {
    const level = Number(params.level);
    const operator = params.operator as ">=" | "<=";
    if (!Number.isFinite(level) || level <= 0) {
      return false;
    }
    return priceLevelCross(previous.close, current.close, level, operator);
  }

  if (type === "price_range") {
    const low = Number(params.low);
    const high = Number(params.high);
    if (!Number.isFinite(low) || !Number.isFinite(high) || low <= 0 || high <= 0 || low >= high) {
      return false;
    }
    return priceRangeBreakout(previous.close, current.close, low, high);
  }

  throw new Error(
    "Alerta custom sin tipo válido (ema, price_ma, rsi, stochastic, price_level o price_range).",
  );
}

function evaluatePreset(
  preset: string,
  params: Record<string, unknown>,
  current: EnrichedBar,
  previous: EnrichedBar,
  enriched: EnrichedBar[],
): boolean {
  switch (preset) {
    case "ema_cross_bull":
      return emaCross(previous.ema_9!, previous.ema_21!, current.ema_9!, current.ema_21!, "up");
    case "ema_cross_bear":
      return emaCross(previous.ema_9!, previous.ema_21!, current.ema_9!, current.ema_21!, "down");
    case "golden_cross":
      return emaCross(previous.ema_50!, previous.ema_200!, current.ema_50!, current.ema_200!, "up");
    case "death_cross":
      return emaCross(previous.ema_50!, previous.ema_200!, current.ema_50!, current.ema_200!, "down");
    case "rsi_oversold":
      return evaluateRsiPreset("rsi_oversold", params, current, enriched);
    case "rsi_overbought":
      return evaluateRsiPreset("rsi_overbought", params, current, enriched);
    case "stoch_oversold":
      return evaluateStochPreset("stoch_oversold", params, enriched);
    case "stoch_overbought":
      return evaluateStochPreset("stoch_overbought", params, enriched);
    case "custom":
      return evaluateCustom(params, enriched);
    default:
      throw new Error(`Preset no soportado: ${preset}`);
  }
}

export function evaluateAlert(alert: Pick<AlertRow, "ticker" | "preset_or_custom" | "params">, bars: OhlcvBar[]): EvaluationResult {
  if (bars.length < 2) {
    throw new Error("Se necesitan al menos 2 velas para evaluar una alerta.");
  }
  const enriched = enrichBars(bars);
  const current = enriched.at(-1)!;
  const previous = enriched.at(-2)!;
  const conditionMet = evaluatePreset(
    alert.preset_or_custom,
    alert.params ?? {},
    current,
    previous,
    enriched,
  );

  return {
    conditionMet,
    candleTimestamp: current.datetime,
    presetOrCustom: alert.preset_or_custom,
    ticker: alert.ticker,
  };
}

File: worker/alert_evaluator.py
"""Evaluación de condiciones de alerta sobre indicadores calculados."""
from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime
from typing import Any

import pandas as pd

from worker.indicator_engine import IndicatorEngine


@dataclass(frozen=True)
class EvaluationResult:
    condition_met: bool
    candle_timestamp: datetime
    preset_or_custom: str
    ticker: str


def _latest_valid_row(df: pd.DataFrame) -> pd.Series:
    if len(df) < 2:
        raise ValueError("Se necesitan al menos 2 velas para evaluar una alerta.")
    return df.iloc[-1]


def _previous_row(df: pd.DataFrame) -> pd.Series:
    return df.iloc[-2]


def _ema_cross(
    prev_fast: float,
    prev_slow: float,
    fast: float,
    slow: float,
    *,
    direction: str,
) -> bool:
    if direction == "up":
        return prev_fast <= prev_slow and fast > slow
    if direction == "down":
        return prev_fast >= prev_slow and fast < slow
    raise ValueError(f"Dirección EMA no soportada: {direction}")


def _rsi_threshold(value: float, threshold: float, operator: str) -> bool:
    if operator == "<":
        return value < threshold
    if operator == ">":
        return value > threshold
    raise ValueError(f"Operador RSI no soportado: {operator}")


class AlertEvaluator:
    """Evalúa si la condición técnica de una alerta se cumple en la vela más reciente."""

    def __init__(self, engine: IndicatorEngine | None = None) -> None:
        self._engine = engine or IndicatorEngine()

    def evaluate(
        self,
        alert: dict[str, Any],
        ohlcv: pd.DataFrame,
    ) -> EvaluationResult:
        enriched = self._engine.enrich(ohlcv)
        current = _latest_valid_row(enriched)
        previous = _previous_row(enriched)
        preset = alert["preset_or_custom"]
        ticker = alert["ticker"]
        candle_ts = pd.Timestamp(current.name).to_pydatetime()

        condition_met = self._evaluate_preset(
            preset=preset,
            params=alert.get("params") or {},
            current=current,
            previous=previous,
            ohlcv=enriched,
        )

        return EvaluationResult(
            condition_met=condition_met,
            candle_timestamp=candle_ts,
            preset_or_custom=preset,
            ticker=ticker,
        )

    def _evaluate_preset(
        self,
        *,
        preset: str,
        params: dict[str, Any],
        current: pd.Series,
        previous: pd.Series,
        ohlcv: pd.DataFrame,
    ) -> bool:
        if preset == "ema_cross_bull":
            return _ema_cross(
                previous["ema_9"],
                previous["ema_21"],
                current["ema_9"],
                current["ema_21"],
                direction="up",
            )
        if preset == "ema_cross_bear":
            return _ema_cross(
                previous["ema_9"],
                previous["ema_21"],
                current["ema_9"],
                current["ema_21"],
                direction="down",
            )
        if preset == "golden_cross":
            return _ema_cross(
                previous["ema_50"],
                previous["ema_200"],
                current["ema_50"],
                current["ema_200"],
                direction="up",
            )
        if preset == "death_cross":
            return _ema_cross(
                previous["ema_50"],
                previous["ema_200"],
                current["ema_50"],
                current["ema_200"],
                direction="down",
            )
        if preset == "rsi_oversold":
            return _rsi_threshold(current["rsi_14"], 30, "<")
        if preset == "rsi_overbought":
            return _rsi_threshold(current["rsi_14"], 70, ">")
        if preset == "custom":
            return self._evaluate_custom(params, current, previous, ohlcv)
        raise ValueError(f"Preset no soportado: {preset}")

    def _evaluate_custom(
        self,
        params: dict[str, Any],
        current: pd.Series,
        previous: pd.Series,
        ohlcv: pd.DataFrame,
    ) -> bool:
        alert_type = params.get("type")
        if alert_type == "ema":
            fast_len = int(params["ema_fast"])
            slow_len = int(params["ema_slow"])
            direction = params["direction"]
            fast_col = f"ema_{fast_len}"
            slow_col = f"ema_{slow_len}"
            if fast_col not in ohlcv.columns or slow_col not in ohlcv.columns:
                ohlcv = ohlcv.copy()
                ohlcv[fast_col] = self._engine.ema(ohlcv, fast_len)
                ohlcv[slow_col] = self._engine.ema(ohlcv, slow_len)
                current = ohlcv.iloc[-1]
                previous = ohlcv.iloc[-2]
            return _ema_cross(
                previous[fast_col],
                previous[slow_col],
                current[fast_col],
                current[slow_col],
                direction=direction,
            )
        if alert_type == "rsi":
            period = int(params.get("period", 14))
            rsi_col = f"rsi_{period}"
            if rsi_col not in ohlcv.columns:
                ohlcv = ohlcv.copy()
                ohlcv[rsi_col] = self._engine.rsi(ohlcv, period)
                current = ohlcv.iloc[-1]
            return _rsi_threshold(
                current[rsi_col],
                float(params["threshold"]),
                params["operator"],
            )
        raise ValueError("Alerta custom sin tipo válido (ema o rsi).")
